### ISSUE TITLE
torch.nonzero returns truncated/wrong results on XPU (torch 2.13.0+xpu)

### ISSUE THREAD

**Issue by ugenc-intel** [intel/torch-xpu-ops#4396]
### Summary:
torch.nonzero returns truncated / wrong results on XPU, elements are silently dropped. The same
tensor on CPU is correct.

### Environment:
PyTorch version: 2.13.0+xpu
TorchVision version: 0.28.0+xpu
Python version: 3.13.7
OS: Ubuntu 25.10
Kernel: 6.17.0-12-generic
XPU available: True
XPU device count: 4
XPU device: Intel(R) Arc(TM) Pro B60 Graphics
XPU driver version: 1.14.37020+3

### Reproduce:
```python
import torch

mask = torch.tensor([0, 1, 0, 1, 0], dtype=torch.bool)

cpu_idx = torch.nonzero(mask)
xpu_idx = torch.nonzero(mask.to("xpu")).cpu()

print(f"torch: {torch.__version__}")
print(f"mask:      {mask.tolist()}")
print(f"cpu index: {cpu_idx.flatten().tolist()}   (expected [1, 3])")
print(f"xpu index: {xpu_idx.flatten().tolist()}")
```

### Versions

PyTorch version: 2.13.0+xpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 25.10 (x86_64)
GCC version: (Ubuntu 15.2.0-4ubuntu4) 15.2.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.42

Python version: 3.13.7 (main, Mar  3 2026, 12:19:54) [GCC 15.2.0] (64-bit runtime)
Python platform: Linux-6.17.0-12-generic-x86_64-with-glibc2.42
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: True
XPU used to build PyTorch: 20260000
Intel GPU driver version:
* intel-opencl-icd:     26.05.37020.3-1~25.10~ppa3
* libze1:       1.27.0-1~25.10~ppa2
Intel GPU models onboard:
* Intel(R) Arc(TM) Pro B60 Graphics
* Intel(R) Arc(TM) Pro B60 Graphics
* Intel(R) Arc(TM) Pro B60 Graphics
* Intel(R) Arc(TM) Pro B60 Graphics
Intel GPU models detected:
* [0] _XpuDeviceProperties(name='Intel(R) Arc(TM) Pro B60 Graphics', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE211, uuid=868011e2-0000-0000-1800-000000000000, driver_version='1.14.37020+3', total_memory=23256MB, local_mem_size=128KB, last_level_cache_size=18432KB, max_compute_units=160, memory_clock_rate=2400MHz, memory_bus_width=64-bit, gpu_eu_count=160, gpu_subslice_count=20, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1, is_integrated_gpu=0)
* [1] _XpuDeviceProperties(name='Intel(R) Arc(TM) Pro B60 Graphics', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE211, uuid=868011e2-0000-0000-3600-000000000000, driver_version='1.14.37020+3', total_memory=23256MB, local_mem_size=128KB, last_level_cache_size=18432KB, max_compute_units=160, memory_clock_rate=2400MHz, memory_bus_width=64-bit, gpu_eu_count=160, gpu_subslice_count=20, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1, is_integrated_gpu=0)
* [2] _XpuDeviceProperties(name='Intel(R) Arc(TM) Pro B60 Graphics', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE211, uuid=868011e2-0000-0000-5400-000000000000, driver_version='1.14.37020+3', total_memory=23256MB, local_mem_size=128KB, last_level_cache_size=18432KB, max_compute_units=160, memory_clock_rate=2400MHz, memory_bus_width=64-bit, gpu_eu_count=160, gpu_subslice_count=20, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1, is_integrated_gpu=0)
* [3] _XpuDeviceProperties(name='Intel(R) Arc(TM) Pro B60 Graphics', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE211, uuid=868011e2-0000-0000-7200-000000000000, driver_version='1.14.37020+3', total_memory=23256MB, local_mem_size=128KB, last_level_cache_size=18432KB, max_compute_units=160, memory_clock_rate=2400MHz, memory_bus_width=64-bit, gpu_eu_count=160, gpu_subslice_count=20, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1, is_integrated_gpu=0)
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  112
On-line CPU(s) list:                     0-111
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) w9-3495X
CPU family:                              6
Model:                                   143
Thread(s) per core:                      2
Core(s) per socket:                      56
Socket(s):                               1
Stepping:                                8
CPU(s) scaling MHz:                      21%
CPU max MHz:                             4800.0000
CPU min MHz:                             800.0000
BogoMIPS:                                3792.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect user_shstk avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.6 MiB (56 instances)
L1i cache:                               1.8 MiB (56 instances)
L2 cache:                                112 MiB (56 instances)
L3 cache:                                105 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-111
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect

**Comment by lchen2331** [comment]
This issue will been fixed in the latest driver version.

Variant | A | B
-- | -- | --
Driver build | 38646 | 37020
nonzero([0,1,0,1,0]) | [1,3] ✅ | [1] ❌

**Comment by chuanqi129** [comment]
> This issue will been fixed in the latest driver version.
> 
> Variant	A	B
> Driver build	38646	37020
> nonzero([0,1,0,1,0])	[1,3] ✅	[1] ❌

Hi @lchen2331 do we need rebuild pytorch with new driver? Or this is just a runtime driver issue?

**Comment by lchen2331** [comment]
> > This issue will been fixed in the latest driver version.
> > Variant	A	B
> > Driver build	38646	37020
> > nonzero([0,1,0,1,0])	[1,3] ✅	[1] ❌
> 
> Hi [@lchen2331](https://github.com/lchen2331) do we need rebuild pytorch with new driver? Or this is just a runtime driver issue?

Hi, no rebuild is needed. This is purely a runtime driver issue.

**Comment by ugenc-intel** [comment]
Confirmed resolved with driver update 38646. For the record, this issue specifically affected cosmos3 model pipeline of diffusers library.

**Comment by Spruill-1** [comment]
Confirming this on driver **26.05.37020.3** (Ubuntu 26.04, kernel 7.0.0-27, torch 2.13.0+xpu) — and adding that it is not limited to the G21/B60 originally reported: it reproduces **identically on Battlemage G31 (Arc Pro B70, `8086:e223`)**, on all three cards we tested (2× B70, 1× B60), stable across weeks of runs. Quantified:

- bool mask with 600 true elements → `torch.nonzero` returns **305–309 rows, varying call-to-call** on the same tensor
- 100 true elements → **0 rows**
- 1,000,000 true elements → **37,824** rows
- meanwhile `mask.sum()`, `x[mask]`, `masked_fill`, `where`, `index_put`, `cumsum` on the same tensors are all correct — the defect is isolated to the `nonzero` result-compaction path

Two notes for anyone landing here:

1. Because everything *around* `nonzero` works, this corrupts silently — e.g. spline/interp code using `nonzero` for bucketing produces plausible-but-wrong results rather than crashing. Worth a loud mention in release notes for 2.13-on-26.05 users.
2. The fix driver (26.22.38646.4) is not currently adoptable on our dual-B70 + dual-B60 host: intel/compute-runtime#922 and intel/compute-runtime#953 are open regressions past 26.05 affecting multi-GPU setups, so we are pinned on 26.05 with the bug live and work around it by avoiding `nonzero`-dependent ops. A UMD/torch-xpu-ops-side mitigation (or a 26.05 backport) would help users stuck in this window.
